### PR TITLE
Dev adjoint strat diag

### PR DESCRIPTION
**Pull request by lxuechen**
The last missing pieces of adjoints:
- Stratonovich diagonal, scalar, additive
- Ito general 

I'll send in another PR later to unify test_strat and test_adjoint.

### CODE REVIEW

File: tests/test_adjoint.py [hunk 1]
```diff
@@ -33,36 +33,30 @@
 device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
 dtype = torch.get_default_dtype()
 
-ito_methods = {'euler': 'ito',
-               'milstein': 'ito',
-               'srk': 'ito'}
-stratonovich_methods = {'midpoint': 'stratonovich',
-                        'log_ode': 'stratonovich'}
```
Comment (patrick-kidger) [inline tests/test_adjoint.py:40]: What's the rationale for removing Euler?

File: tests/test_adjoint.py [hunk 2]
```diff
@@ -33,36 +33,30 @@
...
 def test_adjoint(problem, method, sde_type, noise_type, adaptive):
-    if method == 'euler' and adaptive:
-        return
     if problem is not Ex3 and noise_type == 'additive':
         return
-    # TODO: remove this once we have adjoint implemented for other noise/sde combinations
-    if sde_type == 'stratonovich' and noise_type != 'general':
-        return
     if sde_type == 'ito' and noise_type == 'general':
         return
```
Comment (patrick-kidger) [inline tests/test_adjoint.py:48]: Can this check also be removed now if Ito+general is supported?

File: tests/test_adjoint.py [hunk 3]
```diff
@@ -33,36 +33,30 @@
...
+    if noise_type == "diagonal" and method == "log_ode":
+        return
```
Comment (patrick-kidger) [inline tests/test_adjoint.py:50]: So I get that log_ode is the same as midpoint when the noise is non-general. If that's the reason for this check then all non-general noise_types should trigger this.

That said, I'd prefer to test log-ode anyway. It is a valid combination.

File: tests/test_adjoint.py [hunk 4]
```diff
@@ -33,36 +33,30 @@
...
 
-    d = 1 if noise_type == 'scalar' else 10
+    d = 10
+    m = {"scalar": 1}.get(noise_type, d)  # TODO: Decouple d from m.
```
Comment (patrick-kidger) [inline tests/test_adjoint.py:53]: can this just be `m = 1 if noise_type == 'scalar' else 10`?

File: torchsde/_core/base_sde.py [hunk 5]
```diff
@@ -131,7 +131,7 @@ def dg_ga_jvp_column_sum_v1(self, t, y, a):
 
     def dg_ga_jvp_column_sum_v2(self, t, y, a):
         # Faster, but more memory intensive.
-        requires_grad = torch.is_grad_enabled()
+        requires_grad = torch.is_grad_enabled() and (t.requires_grad or y.requires_grad or a.requires_grad)
```
Comment (patrick-kidger) [inline torchsde/_core/base_sde.py:134]: This probably shuldn't be the case, as per our discussion on my previous PR?

**Review by patrick-kidger — COMMENTED** [review]
I notice that you're including a few of my old commits as well. Can those be dropped?

Overall, very cool to see that support for this is in now.

File: tests/test_adjoint.py (hunk 1 above)
Comment (lxuechen) [reply tests/test_adjoint.py:40]: It's just hard for gradients to match up.
